### ISSUE TITLE
idea: verify the length without needing to read the packet

### ISSUE THREAD

**Issue by dominictarr** [calvinmetcalf/hmac-stream#4]
from what I have read regarding mac then encrypt, i.e
http://www.thoughtcrime.org/blog/the-cryptographic-doom-principle/

it seems the danger is that an attacker can flip bits and get a different length, tricking the verifier into doing that amount of work, and using that work they did to predict the meaning of those bits, thus decrypting a single byte, at least. 

Idea: what if instead of packing each packet with a `length, hmac(header + packet, secret), packet`, you did `length, hash(packet), hmac(header + packet, secret), packet`

This way, the verifier does not need to read the whole packet to see that a bit had been flipped in the length parameter. any bit flipped in the header (length, hash, hmac) would produce the same behavior from the verifier (abort) so all the information that an attacker could get out would be where the packet boundries are (which is more or less public information already)

**Comment by calvinmetcalf** [comment]
this is built around the idea of encrypt *then* mac so the length would not be encrypted, while an attacker could modify the length this would only cause the decrypter do do extra work in the form of calculating the hmac on more of the ciphertext then it needed, so I saw that as more of a DOS issue to deal with via setting a max packet length.

**Comment by dominictarr** [comment]
Yes, that is true... though I have to say I think that crypto primitives (or crypto modules, since this isn't really a primitive, since it's constructed of another primitive, hashes) ought to be air tight with no wiggle room.
The bugs in mac then encrypt often seemed to be that they depended on interactions between the layers
(i.e. non-cryptographically secure padding that is meant to break when bits are flipped)

**Comment by calvinmetcalf** [comment]
Yeah Mac then encrypt probably will work assuming there aren't any other issues , encrypt then mac will catch those issues
